docs(testing): add componentOnReady guidance and a Vue testing page (#4611)
ionic-team/ionic-docs: docs/vue/testing.md

Reviewer: Add 'vue/testing' to the Vue category in sidebars.js, after 'vue/platform' to match where Angular puts it. Without it the page builds but is unreachable, and nothing else in the docs links to it.

---
title: Testing
---

<head>
  <title>Vue Unit and End-to-End Testing for Ionic App Components</title>
  <meta
    name="description"
    content="Vue apps created using Ionic are automatically set up for unit and end-to-end testing. Read to learn more about testing tools for Ionic components."
  />
</head>

This document provides an overview of how to test an application built with `@ionic/vue`. Applications generated with the Ionic CLI are set up for unit testing with [Vitest](https://vitest.dev) and [Vue Test Utils](https://test-utils.vuejs.org), and for end-to-end testing with [Cypress](https://www.cypress.io).

## Unit Testing

### Waiting for Components

> When you need to wait for an Ionic component to render before asserting against its DOM, use the `componentOnReady` helper exported from `@ionic/core`. Do not call `el.componentOnReady()` directly. `@ionic/vue` uses Stencil's custom elements build, where that method does not exist on the element. The helper waits one animation frame instead, giving the component's inner contents a chance to render.

Reviewer: Forgot to mention it in the last review. This should also have an example like the other two frameworks. Consistency is very important for our docs.

Author: Added

---
title: Testing
---

<head>
  <title>Vue Unit and End-to-End Testing for Ionic App Components</title>
  <meta
    name="description"
    content="Vue apps created using Ionic are automatically set up for unit and end-to-end testing. Read to learn more about testing tools for Ionic components."
  />
</head>

> # Testing Ionic Vue

Reviewer: We should be consistent with the other pages and not use the frontmatter title.

```suggestion
```

> This document provides an overview of how to test an application built with `@ionic/vue`. It covers the basics of testing with Vue, as well as the specific tools and libraries developers can use to test their applications.

Reviewer: The old second sentence promised coverage of the specific tools and libraries but the page never named any, so this names the three the Vue starter ships instead.

```suggestion
This document provides an overview of how to test an application built with `@ionic/vue`. Applications generated with the Ionic CLI are set up for unit testing with [Vitest](https://vitest.dev) and [Vue Test Utils](https://test-utils.vuejs.org), and for end-to-end testing with [Cypress](https://www.cypress.io).
```

## Introduction

Testing is an important part of the development process, and it helps to ensure that an application is working as intended.

> 

Reviewer: Remove this section as we don't have it in the other pages. We should be consistent.
```suggestion
```

> ## Best Practices

Reviewer: Let's be consistent with the other frameworks:

```suggestion
## Unit Testing
```

> ### Import `componentOnReady` for standalone projects

Reviewer: ```suggestion
### Waiting for Components
```

> When testing Ionic components, use the exported `componentOnReady` helper from `@ionic/core` instead of calling `el.componentOnReady()` directly. The helper works with both lazy-loaded and custom-element builds, making it more likely the component has finished rendering before making assertions against its rendered DOM or running accessibility tests.

Reviewer: `@ionic/vue` only ships the custom elements build, so the both-builds hedge doesn't apply here; this states the one behavior Vue apps actually get.

```suggestion
When you need to wait for an Ionic component to render before asserting against its DOM, use the `componentOnReady` helper exported from `@ionic/core`. Do not call `el.componentOnReady()` directly. `@ionic/vue` uses Stencil's custom elements build, where that method does not exist on the element. The helper waits one animation frame instead, giving the component's inner contents a chance to render.
```

## Unit Testing

### Waiting for Components

When you need to wait for an Ionic component to render before asserting against its DOM, use the `componentOnReady` helper exported from `@ionic/core`. Do not call `el.componentOnReady()` directly. `@ionic/vue` uses Stencil's custom elements build, where that method does not exist on the element. The helper waits one animation frame instead, giving the component's inner contents a chance to render.

```tsx
import { mount } from '@vue/test-utils';
import { IonApp } from '@ionic/vue';
import { componentOnReady } from '@ionic/core';

import Example from './Example.vue';

const TestComponent = {
  components: { IonApp, Example },
  template: `
    <ion-app>
      <Example />
    </ion-app>
  `,
};

test('renders the submit button', async () => {
  const wrapper = mount(TestComponent);

  const button = wrapper.element.querySelector('ion-button')!;

  await new Promise<void>((resolve) => componentOnReady(button, () => resolve()));

  expect(button.textContent).toContain('Submit');
});

> ```

Reviewer: Dropped the string-template wrapper since it needs Vue's runtime compiler, which isn't in the default Vite build, and switched to `find().element` which is idiomatic VTU and typed non-null.

```suggestion
```ts
import { mount } from '@vue/test-utils';
import { componentOnReady } from '@ionic/core';

import Example from './Example.vue';

test('renders the submit button', async () => {
  const wrapper = mount(Example);

  const button = wrapper.find('ion-button').element;

  await new Promise<void>((resolve) => componentOnReady(button, () => resolve()));

  expect(button.textContent).toContain('Submit');
});
```
```

Reviewer: @Zac-Smucker-Bryan why was the [test](https://github.com/ionic-team/ionic-docs/pull/4611/changes#diff-fbcb242e61459f9f9a3c29c253d30057b5653d8ffa49dbf2ef17b8b772ba4ed7) removed?

Author: @thetaPC When I have been accepting changes from you on github for the tests specifically (including the other test revisions), they are always getting deleted and I add them back (thus why there are so many commits) - it looks like maybe the suggested change didn't have the closing ```? Would github not like that?

Anyway, it was added back in the most recent commit.

Reviewer: Good catch, that's exactly what happened! The suggestion had a code fence inside it and GitHub needs the suggestion fence to be longer than any fence it contains, so it terminated early and applied an empty replacement. My fault, not yours.
